README.md
<center><h1>FACEBOOK CRAWLING BOT</h1></center>

## 1. BACKGROUND
An Python program which helps finding customers for AhaMove sales team.<br>
Automatically look for shipping request posts in pre-defined groups and notify staffs via Telegram.<br>
Staff will be assign using Telegram Username (Ex: @quancao). Therefore, please make sure all staffs have Username on Telegram.<br>
Posts will be simultaneously logged on Google Sheet.

**Skills involved:**
> 1. Python
> 2. Multi-threading
> 3. Selenium
> 4. Google Sheet API
> 5. Telegram API
> 6. SQL

## 2. WHAT THIS DOES
### 2.1. Notify shipping request posts from non-ahamove users
#### 2.1.1. Send notification via Telegram

> **NEW POST** <br>
> **Nội dung:** {post_content} <br>
> **Facebook:** {facebook_profile} <br>
> **Phone:** {phone_detected_in_content} <br>
> **Link:** {post_link} <br>
> {staff}

### 2.1.2. Save posts on Google Sheet for staff
|phone   |time               |content|post                               |profile                     |staff   |note|
|--------|-------------------|-------|-----------------------------------|----------------------------|--------|----|
|84xxxxxx|2019-12-24 14:17:00|ứng 100|https://www.facebook.com/groups/...|https://www.facebook.com/...|@quancao|good|

### 2.1.3. Save posts Google Sheet for log
|phone   |time               |content|post                               |profile                     |
|--------|-------------------|-------|-----------------------------------|----------------------------|
|84xxxxxx|2019-12-24 14:17:00|ứng 100|https://www.facebook.com/groups/...|https://www.facebook.com/...|

### 2.2. Notify shipping request posts from ahamove users via Telegram
**Filter:** Only users in their 40 days old.

> **NEW POST** <br>
> **Nội dung:** {post_content} <br>
> **Facebook:** {facebook_profile} <br>
> **Link:** {post_link} <br>
> #nophone

### 2.3. Store posts
All gathered posts will be stored in database for analytics later on.


CHANGELOG.md
## Version 2.2.1:
- Handle Facebook changed post's timestamp's title attribute.

## Version 2.2.0:
- Storing posts and pushing Telegam messages are now executed consecutively instead of simultaneously.

## Version 2.1.2:
- Handle Facebook changing login screen.

## Version 2.1.1:
- Append site-package path to sys.path in order to run py file w/o venv.
- Set end time for script.
- Handle WebDriverException.

## Version 2.1.0:
- Blacklist keywords: mua, cáng, con chó, bàn, tủ, ghế, bắn.

## Version 2.0.0:
- Separate posts for log and posts for staff. Update posts into 2 different Google Sheets.
    - For staff:
        - Filter AhaMove users.
        - Every profile will only appear once in a day.
    - For log:
        - Keep all posts from AhaMove users and multiple posts from 1 profile.

## Version 1.1.0:
- Slightly increase crawling speed.
- Simultaneously sync to Google Sheet to take note along side of Telegram notification.
- Remove duplicated posts within day.
- Blacklist keyword "bắn"

## Version 1.0.0:
- Beta released

main.py
import pandas as pd
import re
import time
import threading
import requests
import random
from gsheetApi import play_with_gsheet, gsheet_build_service

# Selenium
from selenium import webdriver
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import TimeoutException, NoSuchElementException
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.common.by import By

# Credentials
import accounts

def standby(min=1, max=3):
    time.sleep(random.uniform(min,max))

def open_browser():
    options = webdriver.ChromeOptions()
    options.add_argument("--disable-notifications")
    options.add_argument("--disable-infobars")
    options.add_argument("--mute-audio")

    driver = webdriver.Chrome(options=options)
    return driver

def login_fb(driver, account, password):
    driver.get('https://facebook.com')
    standby()
    driver.find_element_by_id('email').send_keys(account)
    standby()
    driver.find_element_by_id('pass').send_keys(password)
    standby()
    driver.switch_to.active_element.send_keys(Keys.RETURN)


def get_fb_posts(driver, groupId, kwBlacklist):
    if type(kwBlacklist) != 'list':
        kwBlacklist = list(kwBlacklist)
    kwBlacklist = '|'.join(kwBlacklist)

    dataframe = pd.DataFrame(columns=['phone', 'time', 'content', 'post', 'profile'])

    driver.get(f'https://facebook.com/groups/{groupId}?sorting_setting=CHRONOLOGICAL')
    posts = WebDriverWait(driver, 5).until(EC.presence_of_all_elements_located((By.CLASS_NAME, 'userContentWrapper')))
    for p in posts:
        if p.text.find('Vừa xong') != -1 or p.text.find('1 phút') != -1:
            try:
                p.find_element_by_class_name('see_more_link_inner').click()
            except: pass

            content = p.find_element_by_class_name('userContent').text

            if len(re.findall(r'(ứng \d+)|(ship \d+)', content)) != 0 and not re.findall(kwBlacklist, content, re.IGNORECASE):
                try:
                    profile = p.find_element_by_class_name('profileLink').get_attribute('href')
                except:
                    profile = p.find_element_by_link_text(p.find_element_by_class_name('_7tae').text).get_attribute('href')
                if profile.find('profile.php') == -1:
                    profile = profile.split('?')[0]
                else:
                    profile = profile.split('&')[0]

                try:
                    phone = re.search(r'([^0-9]+(0|84|\+84)[-.\s]?\d{1,3}[-.\s]?\d{2,4}[-.\s]?\d{2,4}[-.\s]?\d{2,4})', content).group()
                    phone = re.sub(r'\D+', '', phone)
                    phone = re.sub(r'^0', '84', phone)
                except:
                    phone = None
            
                try:
                    post = p.find_element_by_class_name('_5pcq').get_attribute('href')
                except:
                    break

                post_time = pd.to_datetime(p.find_element_by_class_name('_5ptz').get_attribute('title'))

                dataframe = dataframe.append({'phone':phone, 'time':post_time, 'content':content,
                                              'post':post, 'profile':profile}, ignore_index=True)
            else: continue
    dataframe = dataframe.drop_duplicates(subset='post')
    return dataframe

def push_tele(df, botToken, teleId):
    for i in range(len(df)):
        profile = df.iloc[i]['profile']
        content = df.iloc[i]['content']
        phone = df.iloc[i]['phone']
        post = df.iloc[i]['post']
        staff = df.iloc[i]['staff']

        text = """
<b>NEW POST</b>
<b>Nội dung:</b> {content}
<b>Facebook:</b> {profile}
<b>Phone:</b> {phone}
<b>Link:</b> {post}
{staff}
    """.format(content=content, profile=profile, phone=phone, post=post, staff=staff)

        data = {
            'chat_id': teleId,
            'text': text,
            'parse_mode': 'HTML',
            'disable_web_page_preview': True
        }

        try:
            requests.post("https://api.telegram.org/bot{token}/sendMessage".format(token=botToken), data=data)
        except:
            pass

def gsheet_connection():
    global service1
    global service2
    global service3
    time.sleep(3600)
    service1 = gsheet_build_service()
    service2 = gsheet_build_service()
    service3 = gsheet_build_service()

def get_old_users():
    global oldUsersList
    dfOldUsers = play_with_gsheet(accounts.spreadsheetIdHubspot, 'Sheet1', first_time=True)
    oldUsersList = dfOldUsers.id.tolist()
    time.sleep(3600)

def append_posts_for_staff(spreadsheetId, service):
    global postsToStaff
    while not newDfEvent.isSet():
        gotNewDf = newDfEvent.wait()
        if gotNewDf:
            play_with_gsheet(spreadsheetId, 'Sheet1', dataframe=postsToStaff, method='append', first_time=False, service=service)
            newDfEvent.clear()

def append_posts_for_log(spreadsheetId, service):
    newPosts
    while not newDfEvent.isSet():
        gotNewDf = newDfEvent.wait()
        if gotNewDf:
            play_with_gsheet(spreadsheetId, 'Sheet1', dataframe=newPosts, method='append', first_time=False, service=service)
            newDfEvent.clear()

def assign_staff(df, staffList):
    global num
    staffCol = []
    for _ in range(len(df)):
        staffCol.append(staffList[num])
        if num < len(staffList) - 1:
            num += 1
        else: num = 0
    df = pd.concat([df, pd.Series(staffCol, name='staff')], axis=1, ignore_index=True)
    df.columns = ['phone', 'time', 'content', 'post', 'profile', 'staff']
    df['note'] = ''
    return df

newDfEvent = threading.Event()

spreadsheetIdNoti = accounts.spreadsheetIdNoti
spreadsheetIdLog = accounts.spreadsheetIdLog
spreadsheetIdHubspot = accounts.spreadsheetIdHubspot

groupIdList = accounts.groupIdList
staffList = accounts.staffList
num = 0
kwBlacklist = ['bắn']

postsToStaff = None
newPosts = None

service1 = gsheet_build_service()
service2 = gsheet_build_service()
service3 = gsheet_build_service()

staffPostsThread = threading.Thread(target=append_posts_for_staff, args=(spreadsheetIdNoti, service1,), name='staffPostsThread')
logPostsThread = threading.Thread(target=append_posts_for_log, args=(spreadsheetIdLog, service2,), name='logPostsThread')
oldUsersThread = threading.Thread(target=get_old_users, name='oldUsersThread')
gsheetApiThread = threading.Thread(target=gsheet_connection, name='gsheetApiThread')

staffPostsThread.start()
logPostsThread.start()
oldUsersThread.start()
gsheetApiThread.start()

driver = open_browser()
login_fb(driver, accounts.acc1, accounts.pass1)
standby()

while True:
    for groupId in groupIdList:
        try:
            newPosts = get_fb_posts(driver, groupId, kwBlacklist)
            oldPostsList = play_with_gsheet(spreadsheetIdLog, 'Sheet1', first_time=False, service=service3).post.tolist()
            newPosts = newPosts[~newPosts.post.isin(oldPostsList)]
            oldProfilesList = play_with_gsheet(spreadsheetIdNoti, 'Sheet1!E:E', first_time=False, service=service3).profile.tolist()
            postsToStaff = newPosts[(~newPosts.profile.isin(oldProfilesList)) & (~newPosts.phone.isin(oldUsersList))]
            postsToStaff = assign_staff(postsToStaff.reset_index(drop=True), staffList)
            newDfEvent.set()
            push_tele(postsToStaff, accounts.botToken, accounts.teleIdHan)
        except Exception as err:
            err_text = f"Error: {type(err).__name__}.\n{str(err)}"
            data = {
                'chat_id': '807358017',
                'text': err_text,
                'parse_mode': 'HTML'
            }
            r = requests.post("https://api.telegram.org/bot{token}/sendMessage".format(token=accounts.botToken), data=data)
            continue


main_sgn.py
import sys
sys.path.append(sys.path[0]+'\selenium-fb\Lib\site-packages')

import pandas as pd
import re
import time
import datetime
import schedule
import requests
import random
from gsheetApi import play_with_gsheet, gsheet_build_service

# Selenium
from selenium import webdriver
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import TimeoutException, NoSuchElementException
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.common.by import By

# Credentials
import accounts

def standby(min=1, max=3):
    time.sleep(random.uniform(min,max))

def open_browser():
    options = webdriver.ChromeOptions()
    options.add_argument("--disable-notifications")
    options.add_argument("--disable-infobars")
    options.add_argument("--mute-audio")
    options.add_argument('--no-sandbox')
    options.add_argument('--disable-dev-shm-usage')

    driver = webdriver.Chrome(options=options)
    return driver

def login_fb(driver, account, password):
    driver.get('https://facebook.com')
    standby(3, 5)
    WebDriverWait(driver, 5).until(EC.presence_of_element_located((By.NAME, 'email'))).send_keys(account)
    standby()
    WebDriverWait(driver, 5).until(EC.presence_of_element_located((By.NAME, 'pass'))).send_keys(password)
    standby()
    driver.switch_to.active_element.send_keys(Keys.RETURN)

def get_fb_posts(driver, groupId, kwBlacklist):
    if type(kwBlacklist) != 'list':
        kwBlacklist = list(kwBlacklist)
    kwBlacklist = '|'.join(kwBlacklist)

    dataframe = pd.DataFrame(columns=['phone', 'time', 'content', 'post', 'profile'])

    driver.get(f'https://facebook.com/groups/{groupId}?sorting_setting=CHRONOLOGICAL')
    posts = WebDriverWait(driver, 5).until(EC.presence_of_all_elements_located((By.CLASS_NAME, 'userContentWrapper')))
    for p in posts:
        if p.text.find('Vừa xong') != -1 or p.text.find('1 phút') != -1:
            try:
                p.find_element_by_class_name('see_more_link_inner').click()
            except: pass

            content = p.find_element_by_class_name('userContent').text

            if len(re.findall(r'(ứng \d+)|(ship \d+)', content)) != 0 and not re.findall(kwBlacklist, content, re.IGNORECASE):
                try:
                    profile = p.find_element_by_class_name('profileLink').get_attribute('href')
                except:
                    profile = p.find_element_by_link_text(p.find_element_by_class_name('_7tae').text).get_attribute('href')
                if profile.find('profile.php') == -1:
                    profile = profile.split('?')[0]
                else:
                    profile = profile.split('&')[0]

                try:
                    phone = re.search(r'([^0-9]+(0|84|\+84)[-.\s]?\d{1,3}[-.\s]?\d{2,4}[-.\s]?\d{2,4}[-.\s]?\d{2,4})', content).group()
                    phone = re.sub(r'\D+', '', phone)
                    phone = re.sub(r'^0', '84', phone)
                except:
                    phone = None
            
                try:
                    post = p.find_element_by_class_name('_5pcq').get_attribute('href')
                except:
                    break

                post_time = pd.to_datetime(p.find_element_by_class_name('_5ptz').get_attribute('data-utime'), unit='s') + pd.DateOffset(hour=16)

                dataframe = dataframe.append({'phone':phone, 'time':post_time, 'content':content,
                                              'post':post, 'profile':profile}, ignore_index=True)
            else: continue
    dataframe = dataframe.drop_duplicates(subset='post')
    return dataframe

def push_tele(df, botToken, teleId):
    for i in range(len(df)):
        profile = df.iloc[i]['profile']
        content = df.iloc[i]['content']
        phone = df.iloc[i]['phone']
        post = df.iloc[i]['post']
        staff = df.iloc[i]['staff']

        text = """
<b>NEW POST</b>
<b>Nội dung:</b> {content}
<b>Facebook:</b> {profile}
<b>Phone:</b> {phone}
<b>Link:</b> {post}
{staff}
    """.format(content=content, profile=profile, phone=phone, post=post, staff=staff)

        data = {
            'chat_id': teleId,
            'text': text,
            'parse_mode': 'HTML',
            'disable_web_page_preview': True
        }

        try:
            requests.post("https://api.telegram.org/bot{token}/sendMessage".format(token=botToken), data=data)
        except:
            pass

def get_old_users():
    global oldUsersList
    dfOldUsers = play_with_gsheet(accounts.spreadsheetIdHubspot, 'Sheet1', first_time=True)
    oldUsersList = dfOldUsers.id
    return oldUsersList

def append_posts(spreadsheetId, df, service):
    play_with_gsheet(spreadsheetId, 'Sheet1', dataframe=df, method='append', first_time=False, service=service)

def assign_staff(df, staffList):
    global num
    staffCol = []
    for _ in range(len(df)):
        staffCol.append(staffList[num])
        if num < len(staffList) - 1:
            num += 1
        else: num = 0
    df = pd.concat([df, pd.Series(staffCol, name='staff')], axis=1, ignore_index=True)
    df.columns = ['phone', 'time', 'content', 'post', 'profile', 'staff']
    return df


fb_email = accounts.acc2
fb_pass = accounts.pass2

spreadsheetIdNoti = accounts.spreadsheetIdNotiSgn
spreadsheetIdLog = accounts.spreadsheetIdLogSgn
spreadsheetIdHubspot = accounts.spreadsheetIdHubspot

groupIdList = accounts.groupIdListSgn
staffList = accounts.staffListSgn
num = 0
kwBlacklist = ['bắn', 'mua']

oldUsersList = get_old_users()
schedule.every(30).minutes.do(get_old_users)

driver = open_browser()
login_fb(driver, fb_email, fb_pass)
standby()

while True:
    if datetime.datetime.now().time() >= datetime.time(18, 30) or datetime.datetime.now().time() < datetime.time(9, 29):
        driver.quit()
        sys.exit(0)
    for groupId in groupIdList:
        try:
            schedule.run_pending()
            service = gsheet_build_service()
            newPosts = get_fb_posts(driver, groupId, kwBlacklist)
            oldPostsList = play_with_gsheet(spreadsheetIdLog, 'Sheet1', first_time=False, service=service).post
            newPosts = newPosts[~newPosts.post.isin(oldPostsList)]
            indayPosts = play_with_gsheet(spreadsheetIdNoti, 'Sheet1', first_time=False, service=service)
            oldProfilesList = indayPosts.profile
            oldPhonesList = indayPosts.phone
            postsToStaff = newPosts[(~newPosts.profile.isin(oldProfilesList)) & (~newPosts.phone.isin(oldPhonesList)) &
                                    ((~newPosts.phone.isin(oldUsersList)) | (newPosts.phone.isna()))].drop_duplicates(subset='profile')
            postsToStaff = assign_staff(postsToStaff.reset_index(drop=True), staffList)
            append_posts(accounts.spreadsheetIdNotiSgn, postsToStaff, service)
            push_tele(postsToStaff, accounts.botToken, accounts.teleIdSgn)
            append_posts(accounts.spreadsheetIdLogSgn, newPosts, service)
            del newPosts, postsToStaff, oldPostsList, oldPhonesList
        except Exception as err:
            if type(err).__name__ in ['WebDriverException', 'NoSuchWindowException']:
                try:
                    driver.quit()
                except: pass
                driver = open_browser()
                login_fb(driver, fb_email, fb_pass)
            if type(err).__name__ != 'TimeoutException':
                err_text = f"SGN Error: {type(err).__name__}.\n{str(err)}"
                data = {
                    'chat_id': '807358017',
                    'text': err_text,
                    'parse_mode': 'HTML'
                }
                r = requests.post("https://api.telegram.org/bot{token}/sendMessage".format(token=accounts.botToken), data=data)
            continue

gsheetApi.py
import pandas as pd
import pickle
from googleapiclient.discovery import build
from google_auth_oauthlib.flow import InstalledAppFlow
from google.auth.transport.requests import Request

import requests
import accounts

def gsheet_build_service(token_refresh=True):
    SCOPES = ['https://www.googleapis.com/auth/spreadsheets']
    with open('token.pickle', 'rb') as token:
        credentials = pickle.load(token)
    if not credentials or not credentials.valid:
        if credentials and credentials.expired and credentials.refresh_token and token_refresh:
            credentials.refresh(Request())
        else:
            flow = InstalledAppFlow.from_client_secrets_file('credentials.json', SCOPES)
            credentials = flow.run_local_server(port=0)
        with open('token.pickle', 'wb') as token:
            pickle.dump(credentials, token)
    service = build('sheets', 'v4', credentials=credentials)
    return service


def play_with_gsheet(spreadsheetId=None, _range=None, dataframe=None, method='read', first_time=True, service=None):
    """
    method: {'read', 'write', 'clear', 'append'}
    """
    
    if first_time:
        service = gsheet_build_service()

    if method == 'read':
        result = service.spreadsheets().values().get(spreadsheetId=spreadsheetId, range=_range).execute()
        values = result.get('values', [])
        df = pd.DataFrame(values)
        df = df.iloc[1:].rename(columns=df.iloc[0])
        return df
    
    if method == 'write':
        values = [dataframe.columns.values.astype(str).tolist()] + dataframe.astype(str).values.tolist()
        data = [
        {
            'range': _range,
            'values': values
        }
        ]
        body = {
            'valueInputOption':'RAW',
            'data':data
        }
        
        service.spreadsheets().values().batchUpdate(spreadsheetId=spreadsheetId, body=body).execute()
            
    if method == 'clear':
        body = {
            'ranges':_range
        }
        
        service.spreadsheets().values().batchClear(spreadsheetId=spreadsheetId, body=body).execute()
            
    if method == 'append' and len(dataframe) > 0:
        body = {
            'values': dataframe.astype(str).values.tolist()
        }
        
        service.spreadsheets().values().append(spreadsheetId=spreadsheetId, range=_range,
                                                        valueInputOption='RAW', insertDataOption='INSERT_ROWS',
                                                        body=body).execute()

main_han.py
import sys
sys.path.append(sys.path[0]+'\selenium-fb\Lib\site-packages')

import pandas as pd
import re
import time
import datetime
import schedule
import requests
import random
from gsheetApi import play_with_gsheet, gsheet_build_service

# Selenium 
from selenium import webdriver
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import TimeoutException, NoSuchElementException
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.common.by import By

# Credentials
import accounts

def standby(min=1, max=3):
    time.sleep(random.uniform(min,max))

def open_browser():
    options = webdriver.ChromeOptions()
    options.add_argument("--disable-notifications")
    options.add_argument("--disable-infobars")
    options.add_argument("--mute-audio")
    options.add_argument('--no-sandbox')
    options.add_argument('--disable-dev-shm-usage')

    driver = webdriver.Chrome(options=options)
    return driver

def login_fb(driver, account, password):
    driver.get('https://facebook.com')
    standby(3, 5)
    WebDriverWait(driver, 5).until(EC.presence_of_element_located((By.NAME, 'email'))).send_keys(account)
    standby()
    WebDriverWait(driver, 5).until(EC.presence_of_element_located((By.NAME, 'pass'))).send_keys(password)
    standby()
    driver.switch_to.active_element.send_keys(Keys.RETURN)

def get_fb_posts(driver, groupId, kwBlacklist):
    if type(kwBlacklist) != 'list':
        kwBlacklist = list(kwBlacklist)
    kwBlacklist = '|'.join(kwBlacklist)

    dataframe = pd.DataFrame(columns=['phone', 'time', 'content', 'post', 'profile'])

    driver.get(f'https://facebook.com/groups/{groupId}?sorting_setting=CHRONOLOGICAL')
    posts = WebDriverWait(driver, 5).until(EC.presence_of_all_elements_located((By.CLASS_NAME, 'userContentWrapper')))
    for p in posts:
        if p.text.find('Vừa xong') != -1 or p.text.find('1 phút') != -1:
            try:
                p.find_element_by_class_name('see_more_link_inner').click()
            except: pass

            content = p.find_element_by_class_name('userContent').text

            if len(re.findall(r'(ứng \d+)|(ship \d+)', content)) != 0 and not re.findall(kwBlacklist, content, re.IGNORECASE):
                try:
                    profile = p.find_element_by_class_name('profileLink').get_attribute('href')
                except:
                    profile = p.find_element_by_link_text(p.find_element_by_class_name('_7tae').text).get_attribute('href')
                if profile.find('profile.php') == -1:
                    profile = profile.split('?')[0]
                else:
                    profile = profile.split('&')[0]

                try:
                    phone = re.search(r'([^0-9]+(0|84|\+84)[-.\s]?\d{1,3}[-.\s]?\d{2,4}[-.\s]?\d{2,4}[-.\s]?\d{2,4})', content).group()
                    phone = re.sub(r'\D+', '', phone)
                    phone = re.sub(r'^0', '84', phone)
                except:
                    phone = None
            
                try:
                    post = p.find_element_by_class_name('_5pcq').get_attribute('href')
                except:
                    break

                post_time = pd.to_datetime(p.find_element_by_class_name('_5ptz').get_attribute('data-utime'), unit='s') + pd.DateOffset(hour=16)

                dataframe = dataframe.append({'phone':phone, 'time':post_time, 'content':content,
                                              'post':post, 'profile':profile}, ignore_index=True)
            else: continue
    dataframe = dataframe.drop_duplicates(subset='post')
    return dataframe

def push_tele(df, botToken, teleId):
    for i in range(len(df)):
        profile = df.iloc[i]['profile']
        content = df.iloc[i]['content']
        phone = df.iloc[i]['phone']
        post = df.iloc[i]['post']
        staff = df.iloc[i]['staff']

        text = """
<b>NEW POST</b>
<b>Nội dung:</b> {content}
<b>Facebook:</b> {profile}
<b>Phone:</b> {phone}
<b>Link:</b> {post}
{staff}
    """.format(content=content, profile=profile, phone=phone, post=post, staff=staff)

        data = {
            'chat_id': teleId,
            'text': text,
            'parse_mode': 'HTML',
            'disable_web_page_preview': True
        }

        try:
            requests.post("https://api.telegram.org/bot{token}/sendMessage".format(token=botToken), data=data)
        except:
            pass

def get_old_users():
    global oldUsersList
    dfOldUsers = play_with_gsheet(accounts.spreadsheetIdHubspot, 'Sheet1', first_time=True)
    oldUsersList = dfOldUsers.id
    return oldUsersList

def append_posts(spreadsheetId, df, service):
    play_with_gsheet(spreadsheetId, 'Sheet1', dataframe=df, method='append', first_time=False, service=service)

def assign_staff(df, staffList):
    global num
    staffCol = []
    for _ in range(len(df)):
        staffCol.append(staffList[num])
        if num < len(staffList) - 1:
            num += 1
        else: num = 0
    df = pd.concat([df, pd.Series(staffCol, name='staff')], axis=1, ignore_index=True)
    df.columns = ['phone', 'time', 'content', 'post', 'profile', 'staff']
    return df

fb_email = accounts.acc1
fb_pass = accounts.pass1

spreadsheetIdNoti = accounts.spreadsheetIdNotiHan
spreadsheetIdLog = accounts.spreadsheetIdLogHan
spreadsheetIdHubspot = accounts.spreadsheetIdHubspot

groupIdList = accounts.groupIdListHan
staffList = accounts.staffListHan
num = 0
kwBlacklist = ['bắn', 'mua', 'cáng', 'con chó', 'bàn', 'tủ', 'ghế',
                'ck', 'momo', 'đệm', 'bank', 'banking', 'tuyển dụng']

oldUsersList = get_old_users()
schedule.every(30).minutes.do(get_old_users)

driver = open_browser()
login_fb(driver, fb_email, fb_pass)
standby()

while True:
    if datetime.datetime.now().time() >= datetime.time(18, 30) or datetime.datetime.now().time() < datetime.time(9, 29):
        driver.quit()
        sys.exit(0)
    for groupId in groupIdList:
        try:
            schedule.run_pending()
            service = gsheet_build_service()
            newPosts = get_fb_posts(driver, groupId, kwBlacklist)
            oldPostsList = play_with_gsheet(spreadsheetIdLog, 'Sheet1', first_time=False, service=service).post
            newPosts = newPosts[~newPosts.post.isin(oldPostsList)]
            inDayPosts = play_with_gsheet(spreadsheetIdNoti, 'Sheet1', first_time=False, service=service)
            oldProfilesList = inDayPosts.profile
            oldPhonesList = inDayPosts.phone
            postsToStaff = newPosts[(~newPosts.profile.isin(oldProfilesList)) & (~newPosts.phone.isin(oldPhonesList)) &
                                    (~newPosts.phone.isin(oldUsersList))].drop_duplicates(subset='profile')
            postsToStaff = assign_staff(postsToStaff.reset_index(drop=True), staffList)
            append_posts(accounts.spreadsheetIdNotiHan, postsToStaff, service)
            push_tele(postsToStaff, accounts.botToken, accounts.teleIdHan)
            append_posts(accounts.spreadsheetIdLogHan, newPosts, service)
        except Exception as err:
            if type(err).__name__ in ['WebDriverException', 'NoSuchWindowException']:
                try:
                    driver.quit()
                except: pass
                driver = open_browser()
                login_fb(driver, fb_email, fb_pass)
            if type(err).__name__ != 'TimeoutException':
                err_text = f"HAN Error: {type(err).__name__}.\n{str(err)}"
                data = {
                    'chat_id': '807358017',
                    'text': err_text,
                    'parse_mode': 'HTML'
                }
                r = requests.post("https://api.telegram.org/bot{token}/sendMessage".format(token=accounts.botToken), data=data)
            continue